pilot/pkg/config/mcp/controller.go

```diff
@@ -0,0 +1,228 @@
+)
+
+var errUnsupported = errors.New("this operation is not supported by mcp controller")
+
+// CoreDataModel is a combined interface for ConfigStoreCache
+// MCP Updater and ServiceDiscovery
+type CoreDataModel interface {
+	model.ConfigStoreCache
+	mcpclient.Updater
+}
+
+// Controller is a temporary storage for the changes received
+// via MCP server
+type Controller struct {
+	configStore              map[string]map[string]*sync.Map
+	eventCh                  chan func(model.Config, model.Event)
+	descriptorsByMessageName map[string]model.ProtoSchema
+}
+
+// NewController provides a new CoreDataModel controller
+func NewController() CoreDataModel {
+	descriptorsByMessageName := make(map[string]model.ProtoSchema, len(model.IstioConfigTypes))
+	for _, config := range model.IstioConfigTypes {
+		descriptorsByMessageName[config.MessageName] = config
+	}
+
+	// Remove this when https://github.com/istio/istio/issues/7947 is done
+	configStore := make(map[string]map[string]*sync.Map)
+	for _, typ := range model.IstioConfigTypes.Types() {
+		configStore[typ] = make(map[string]*sync.Map)
+	}
+	return &Controller{
+		configStore:              configStore,
+		eventCh:                  make(chan func(model.Config, model.Event)),
+		descriptorsByMessageName: descriptorsByMessageName,
+	}
+}
+
+// ConfigDescriptor returns all the ConfigDescriptors that this
+// controller is responsible for
+func (c *Controller) ConfigDescriptor() model.ConfigDescriptor {
+	return model.IstioConfigTypes
+}
+
+// List returns all the config that is stored by type and namespace
+func (c *Controller) List(typ, namespace string) ([]model.Config, error) {
+	_, ok := c.ConfigDescriptor().GetByType(typ)
+	if !ok {
+		return nil, fmt.Errorf("List: unknown type %s", typ)
+	}
+	data, exists := c.configStore[typ]
+	if !exists {
+		log.Infof("List: config not found for the type %s", typ)
+		return nil, nil
+	}
+	out := make([]model.Config, 0, len(c.configStore[typ]))
+	if namespace == "" {
+		for _, ns := range data {
+			ns.Range(func(key, value interface{}) bool {
+				out = append(out, value.(model.Config))
+				return true
+			})
+		}
+	} else {
+		ns, exists := data[namespace]
+		if !exists {
+			return nil, nil
+		}
+		ns.Range(func(key, value interface{}) bool {
+			out = append(out, value.(model.Config))
+			return true
+		})
+	}
+	return out, nil
+}
+
+// Apply receives changes from MCP server and creates the
+// corresponding config
+func (c *Controller) Apply(change *mcpclient.Change) error {
+	for _, obj := range change.Objects {
+		descriptor, ok := c.descriptorsByMessageName[change.MessageName]
+		if !ok {
+			return fmt.Errorf("Apply: type not supported %s", change.MessageName)
+		}
+
+		conf := model.Config{
+			ConfigMeta: model.ConfigMeta{
+				Type:              descriptor.Type,
+				Group:             descriptor.Group,
+				Version:           descriptor.Version,
+				Name:              obj.Metadata.Name,
+				CreationTimestamp: meta_v1.Time{},
```

Comment:
where is the namespace?

Reply:
MCP envelope's metadata doesn't have namespace (yet?). One thought was that MCP server would resolve namespace to FQDN before pushing to clients. This essentially encodes the namespace into the name itself, e.g. <name>/<namespace>.  This is what we are doing today with Galley and Mixer is temporarily splitting up the meta.name to get the namespace (see https://github.com/istio/istio/blob/6acd9fd9ed1fa3bf7f77940ed0a7a5ee24fe30a3/mixer/pkg/config/mcp/conversion.go#L126-L142)

Reply:
Does that mean we have to do the string parsing to split a resource name into name.namespace? I dont mind, but the syntax has to be explicit. I asked about namespace because we use the ConfigMeta namespace to convert short names in virtual services into FQDNs. That logic is pilot specific, so not sure if it can be offloaded to galley.

Reply:
We can consider adding `namespace` to the MCP envelope's metadata if `namespace` is going to continue to be part of the core pilot data model after [decomposition](https://goo.gl/qAsEQx). Otherwise, the out-of-process platform specific service registry layer could normalize FQDN.